Added account age to user info window

poixen opened the pull request:
Added the ability to see the account age of registered users. Changes are both in server and client. The server will now send how long the account has been active in seconds. The client will then use this to calculate the account age and display it in the user details window.

![account age](https://cloud.githubusercontent.com/assets/2134793/5672889/772dcc30-9795-11e4-98f8-e5f3483a7d6e.png)

Daenyth commented on cockatrice/src/userinfobox.cpp:
```diff
@@ -26,6 +28,8 @@ UserInfoBox::UserInfoBox(AbstractClient *_client, bool _fullInfo, QWidget *paren
     userLevelLabel2 = new QLabel;
     userLevelLabel3 = new QLabel;
+    accountAgeLebel1 = new QLabel;
+    accountAgeLabel2 = new QLabel;
```
Can you convert the `new` calls to stack allocation instead? https://github.com/Cockatrice/Cockatrice/blob/master/CONTRIBUTING.md#memory-management

Daenyth commented on cockatrice/src/userinfobox.cpp:
```diff
@@ -79,6 +86,16 @@ void UserInfoBox::updateInfo(const ServerInfo_User &user)
+    QString accountAgeString = tr("Unregistered user");
+    if (userLevel.testFlag(ServerInfo_User::IsAdmin) || userLevel.testFlag(ServerInfo_User::IsModerator) || userLevel.testFlag(ServerInfo_User::IsRegistered)) {
+        QDateTime regDate = QDateTime::fromString(QString::fromStdString(user.registrationdate()), "yyyy-MM-dd hh:mm:ss");
+        qint64 accountAgeInDays = regDate.daysTo(QDateTime::currentDateTime());
```
Could there be a time zone discrepancy here? The server probably sends back UTC time, does currentDateTime give you UTC or local?

Daenyth commented on cockatrice/src/userinfobox.cpp:
```diff
@@ -79,6 +86,16 @@ void UserInfoBox::updateInfo(const ServerInfo_User &user)
+        qint64 accountAgeInDays = regDate.daysTo(QDateTime::currentDateTime());
+        accountAgeString = QString::number(accountAgeInDays);
+        accountAgeString.append(" ");
+        accountAgeString.append(tr("Days"));
```
I think a more fine grained fuzzy time would be appropriate. My account would probably have something silly like 1487 days, which isn't terribly useful.

ZeldaZach replied:
Maybe it can convert that to X years, X months, X days?

Daenyth commented on servatrice/src/servatrice_database_interface.cpp:
```diff
@@ -283,6 +283,11 @@ ServerInfo_User Servatrice_DatabaseInterface::evalUserQueryResult(const QSqlQuer
 		result.set_real_name(realName.toStdString());
+
+    const QString registrationDate = query.value(7).toString();
+    qDebug() << "Registration Date: " << registrationDate; // todo remove
```
Remove?

Daenyth commented:
You note that the column type has to be changed, can you modify the sql bootstrap script to create the table with the right type?

Does this require a change in how the php UI inserts user accounts?

woogerboy21 commented:
Good question , i dont think so since we just do a basic insert call and dont define type cast when making the insert but we can test it.

ctrlaltca commented on servatrice/src/servatrice_database_interface.cpp:
```diff
@@ -283,6 +283,11 @@ ServerInfo_User Servatrice_DatabaseInterface::evalUserQueryResult(const QSqlQuer
 	if (!realName.isEmpty())
 		result.set_real_name(realName.toStdString());
+
+    const QString registrationDate = query.value(7).toString();
```
I'm quite skeptical of storing dates in a char(20) field, since it will force us to use CAST() when we need to make a mathematical computation on the date in the db using sql.
I suppose we could achieve the same result using: 
```
const QString registrationDate = query.value(7).toDateTime().toString(Qt::ISODate);
```
As an alternative, the query could return the value of ``` DATEDIFF(NOW(), registrationDate) ``` as the number of days  in an integer field, avoiding any date calculation clientside an thus any timezone problem.

poixen commented:
I have taken all comments into consideration and have done the following:

DB no longer needs to be changed
Thanks to @ctrlaltca comment I have removed the need for the DB to be changed. The server will now convert the mysql datetime into a local QDateTime, from there I calculate the seconds the account has been active and send that back to the client as a long.

Client now shows years and days
The cleint will now show the age of the account in years and days.

Removed heap alloc
Thanks to comments by @Daenyth I have remove some heap alloc, in favor of the stack.

ctrlaltca commented:
Much better! Code seems fine now, tested and working as expected. :+1: 
Only one thing: on servers not having this patch, the cockatrice client will show "Account age: 0 days" for every user. Maybe if ``` user.accountage_secs() == 0 ``` we could just show "unknown" instead?

ZeldaZach commented:
I'd prefer "Signed up today" over "unknown" @ctrlaltca

ctrlaltca commented:
The server is returning the account age in seconds. A value of zero seconds could mean the server is returning no value at all (missing this patch) or the account has been registered zero seconds ago.. I assume nobody will ever be that fast!

ZeldaZach commented:
Can we just have it round up?[

poixen commented:
@ctrlaltca good points, added a check for 0 secs, if this happens then the account age is "unknown".

ctrlaltca commented:
Tested again, fine for me :+1:

ZeldaZach commented:
I tested and had issues with it (on cockatricetesting server)

When I registered an account and it had the signup date as `2015-01-10 22:43:33` the following occurred; 

![screenshot 2015-01-10 17 51 45](https://cloud.githubusercontent.com/assets/7460172/5693619/353db480-98f2-11e4-833f-6acfe1f9e60f.png)

I also tried altering your date in the DB and set it to 1 year back (I changed it to `2013-10-10 17:10:10`) from when it was listed as, and it continues to say 92 days 
![screenshot 2015-01-10 17 58 57](https://cloud.githubusercontent.com/assets/7460172/5693622/5d38536e-98f2-11e4-9943-06f3821f8428.png)

ZeldaZach commented:
Unwrapping the math fixes it :+1: Good to go on my end

poixen commented:
Removed integer division.
